### PR TITLE
fix: Proxy closing listener with 0 connections left

### PR DESCRIPTION
**Pull request by elimt**
A bug was introduced in `0.15.0` where sessions with max connection counts were closed when the connections left hit 0. This issue was highlighted by a user: https://github.com/hashicorp/boundary/issues/4364. What was happening was when we have zero connections left the proxy code was closing the listener which closed its connections

To fix this:
- `boundary connect  -target-id=xx -host-id=xx` command with no connections left will now wait for all connections to be closed and before exiting the session just like it did in `0.14.3`
- As soon as there are no more connections left, the listener will close to stop accepting new connections
- Fix issue where the websocket connection was throwing closing errors during the session teardown.

### CODE REVIEW

File: internal/cmd/commands/connect/funcs.go [hunk 1]
```diff
@@ -114,7 +114,8 @@ func generateConnectionInfoTableOutput(in ConnectionInfo) string {
 	var ret []string
 
 	nonAttributeMap := map[string]any{
-		"Connections Left": in.ConnectionsLeft,
+		"Connections Left":  in.ConnectionsLeft,
+		"Connections Count": in.ConnectionsCount,
```
Comment (elimt) [inline internal/cmd/commands/connect/funcs.go:118]: This is completely optional. It make it easy for me to debug stuff so I left it in. Happy to take it out

**Review by moduli — COMMENTED** [review]
One of the tests is reporting some different behavior.
 
If I run something like this on a terminal
```
cmdChan <- e2e.RunCommand(ctxCancel, "boundary",
	e2e.WithArgs(
		"connect",
		"-target-id", newTargetId,
		"-listen-port", port,
		"-format", "json",
	),
)
```

It looks like it's printing out more information
```
{"address":"127.0.0.1","port":12345,"protocol":"tcp","expiration":"2024-02-10T02:29:27.429963Z","connection_limit":0,"session_id":"s_TkXrVSlzWV"}
        {"connections_left":-1,"connections_count":1}
        {"connections_left":-1,"connections_count":0}
        {"connections_left":-1,"connections_count":1}
        {"connections_left":-1,"connections_count":0}
        {"connections_left":-1,"connections_count":1}
        {"connections_left":-1,"connections_count":0}
        {"connections_left":-1,"connections_count":1}
        {"connections_left":-1,"connections_count":0}
```
The test is only expecting one json blob to come out of the output of the command and is thus failing to parse it. Not sure if that extra `connections_left, connections_count` output is intended.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,19 @@ Canonical reference for changes, improvements, and bugfixes for Boundary.
 
 ## Next
 
+### Bug Fixes
+
+* cli: Update proxy listener to not close when the number of connections left 
+  for the session is zero. The listener will refuse new connections when the 
+  number of connections left is zero but existing connections will be active. 
+  This fixes a CLI client issue where sessions with max connection count 
+  configured were closed when the number connections left hit 0.
```
Comment (jefferai) [inline CHANGELOG.md:13]: Missing "of" (number of connections)

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,19 @@ Canonical reference for changes, improvements, and bugfixes for Boundary.
...
+  ([Issue](https://github.com/hashicorp/boundary/issues/4364),
+  ([PR](https://github.com/hashicorp/boundary/pull/4389)) 
+* Safely close websocket connection during session teardown. Fix issue where
```
Comment (jefferai) [inline CHANGELOG.md:16]: It was safe before, this makes it sound unsafe 😀 I'd just say that it fixes a spurious error or so

**Review by jefferai — APPROVED** [review]

**Pull request merged**